agent/src/beerocks/slave/agent_ucc_listener.h

```diff
@@ -0,0 +1,49 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef __AGENT_UCC_LISTENER_H__
+#define __AGENT_UCC_LISTENER_H__
+
+#include <beerocks/bcl/beerocks_config_file.h>
+#include <beerocks/bcl/beerocks_ucc_listener.h>
+
+#include <mutex>
+
+namespace beerocks {
+class agent_ucc_listener : public beerocks_ucc_listener {
+public:
+    agent_ucc_listener(uint16_t port, const std::string &vendor, const std::string &model,
+                       const std::string &bridge_iface, Socket **controller_sd);
+    ~agent_ucc_listener(){};
+
+    bool init() override;
+
+    void lock() override { mutex.lock(); }
+    void unlock() override { mutex.unlock(); }
+
+private:
+    std::string fill_version_reply_string() override;
+    void clear_configuration() override;
+    Socket *get_dev_send_1905_destination_socket(const std::string &dest_alid) override;
+    std::shared_ptr<uint8_t> get_buffer_filled_with_cmdu() override;
+    bool send_cmdu_to_destination(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx) override;
+    bool handle_dev_set_config(std::unordered_map<std::string, std::string> &params,
+                               std::string &err_string) override;
+
+    const std::string &m_vendor;
+    const std::string &m_model;
+    const std::string &m_bridge_iface;
+    std::string m_bridge_mac;
+    Socket **const m_controller_sd;
```

Comment:
Oh, now I see it.... Why is it `**`?? Why not a simple pointer or shared_ptr?

Reply:
It is pointer to pointer since the son_slave is manging the controller socket pointer and it can modify it. That way it is promised that if the son_slave will change it, the ucc_listener would "know" about it.

Reply:
This should be a shared_ptr then, and the son_slave should notify the listener on changes to the socket instead of relying on the pointer to be changed "under your feet".

Reply:
I don't agree with you. IMO this is the most clean/transparent way to do it. The design of the ucc_listener is to have a shared memory with another thread. On the master thread is easy since it has the controller database which holds the relevant sockets is on another file. Here it is more problematic since the socket is on the son_slave and I can't create a circular #include.
As well this is not the purpose of shared_ptr.
shared_ptr purourse is ti hold and not free a memory block as long as it is used. This is not the case here. Even if I turned the pointer to shared_ptr I would still have to pass it as a pointer or reference to shared_ptr to the ucc listener.
As far as I can see these are the alternative solutions:
1. Instead of passing the pointer to the controller socket, passing a callback function (lambda) which will give the current controller socket. The only drawback in this solution is that every time the ucc listener will need the controller socket, it will have to call to the callback function.
2. Create a public function on the ucc_listener that updates the controller socket.
This is a worse idea since every place the son_slave will want to change the controller socket, it will have to call this function. And of course, if a developer will want to change the controller socket in the future, it will have to remember to call this function. This can easily can be forgotten and can lead eventually to bugs.

Reply:
> I don't agree with you. IMO this is the most clean/transparent way to do it. The design of the ucc_listener is to have a shared memory with another thread. On the master thread is easy since it has the controller database which holds the relevant sockets is on another file. Here it is more problematic since the socket is on the son_slave and I can't create a circular #include.

Can't you just use `extern`? no need for circular `#include`.

Reply:
I'll tell you what, When I will create the unified agent process, I'll create for it a database class like the controller has. That way the agent class will be much cleaner.